Mathlib/RingTheory/Smooth/Kaehler.lean

```diff
@@ -224,3 +238,198 @@ def retractionKerToTensorEquivSection :
+  let e₁ : (RingHom.ker (algebraMap P S)).Cotangent ≃ₗ[P] (RingHom.ker (algebraMap P' S)) := by
+    refine (Ideal.cotangentEquivIdeal _).trans ((LinearEquiv.ofEq _ _
+      (IsScalarTower.toAlgHom R P S).ker_kerSquareLift.symm).restrictScalars P)
+  let e₂ : S ⊗[P'] Ω[P'⁄R] ≃ₗ[P] S ⊗[P] Ω[P⁄R] :=
+    (tensorKaehlerQuotKerSqEquiv R P S).restrictScalars P
+  have H : kerCotangentToTensor R P S =
+      e₂.toLinearMap ∘ₗ (kerToTensor R P' S ).restrictScalars P ∘ₗ e₁.toLinearMap := by
+    ext x
+    obtain ⟨x, rfl⟩ := Ideal.toCotangent_surjective _ x
+    exact (tensorKaehlerQuotKerSqEquiv_tmul_D 1 x.1).symm
+  refine Equiv.trans ?_ (retractionKerToTensorEquivSection (R := R) h₂ h₁)
+  refine ⟨fun ⟨l, hl⟩ ↦ ⟨⟨(e₁.toLinearMap ∘ₗ l ∘ₗ e₂.toLinearMap).toAddMonoidHom, ?_⟩, ?_⟩,
+    fun ⟨l, hl⟩ ↦ ⟨e₁.symm.toLinearMap ∘ₗ l.restrictScalars P ∘ₗ e₂.symm.toLinearMap, ?_⟩, ?_, ?_⟩
+  · rintro x y
+    obtain ⟨x, rfl⟩ := Ideal.Quotient.mk_surjective x
+    simp only [← Ideal.Quotient.algebraMap_eq, IsScalarTower.algebraMap_smul]
+    exact (e₁.toLinearMap ∘ₗ l ∘ₗ e₂.toLinearMap).map_smul x y
+  · ext1 x
+    rw [H] at hl
+    obtain ⟨x, rfl⟩ := e₁.surjective x
+    exact DFunLike.congr_arg e₁ (LinearMap.congr_fun hl x)
+  · ext x
+    rw [H]
+    apply e₁.injective
+    simp only [LinearMap.coe_comp, LinearEquiv.coe_coe, LinearMap.coe_restrictScalars,
+      Function.comp_apply, LinearEquiv.symm_apply_apply, LinearMap.id_coe, id_eq,
+      LinearEquiv.apply_symm_apply]
+    exact LinearMap.congr_fun hl (e₁ x)
+  · intro f
+    ext x
+    simp only [AlgebraTensorModule.curry_apply, Derivation.coe_comp, LinearMap.coe_comp,
+      LinearMap.coe_restrictScalars, Derivation.coeFn_coe, Function.comp_apply, curry_apply,
+      LinearEquiv.coe_coe, LinearMap.coe_mk, AddHom.coe_coe, LinearMap.toAddMonoidHom_coe,
+      LinearEquiv.apply_symm_apply, LinearEquiv.symm_apply_apply]
+  · intro f
+    ext x
+    simp only [AlgebraTensorModule.curry_apply, Derivation.coe_comp, LinearMap.coe_comp,
+      LinearMap.coe_restrictScalars, Derivation.coeFn_coe, Function.comp_apply, curry_apply,
+      LinearMap.coe_mk, AddHom.coe_coe, LinearMap.toAddMonoidHom_coe, LinearEquiv.coe_coe,
+      LinearEquiv.symm_apply_apply, LinearEquiv.apply_symm_apply]
+
+variable [Algebra.FormallySmooth R P]
+
+include hf in
+/--
+Given a formally smooth `R`-algebra `P` and a surjective algebra homomorphism `f : P →ₐ[R] S`
+with kernel `I` (typically a presentation `R[X] → S`),
+`S` is formally smooth iff the `P`-linear map `I/I² → B ⊗[A] Ω[A⁄R]` is split injective.
+-/
+theorem Algebra.FormallySmooth.iff_split_injection :
+    Algebra.FormallySmooth R S ↔ ∃ l, l ∘ₗ (kerCotangentToTensor R P S) = LinearMap.id := by
+  have := (retractionEquivSectionKerCotangentToTensor (R := R) hf).nonempty_congr
+  simp only [nonempty_subtype] at this
+  rw [this, ← Algebra.FormallySmooth.iff_split_surjection _ hf]
+
+include hf in
+theorem Algebra.FormallySmooth.iff_injective_and_split :
```

Comment:
```suggestion
include hf in
/--
Given a formally smooth `R`-algebra `P` and a surjective algebra homomorphism `f : P →ₐ[R] S`
with kernel `I` (typically a presentation `R[X] → S`),
then `S` is formally smooth iff `I/I² → S ⊗[P] Ω[S⁄R]` is injective and
`S ⊗[P] Ω[P⁄R] → Ω[S⁄R]` is split injective.
-/
theorem Algebra.FormallySmooth.iff_injective_and_split :
```
I find these maps hard to visualize from the name only, hence more docstrings help.